Extensive matching

SvobodaJakub opened the pull request:
Hi @julian-klode, I like your dns66 quite a lot and had an idea that it might be possible to make the domain-matching algorithm more extensive.

Some lists (such as Badd Boyz Hosts) contain only second-level domains but it's desirable to block any access to them - even to sub-domains.

This is a rough quick idea sketch not intended for merge, described in the commit and described in the code. Works fine for me so far. My aim is to get your opinion.

What do you think about it? Is it worthwile? Something I apparently didn't consider? Would there be a better way?

Some of my ideas for next actions:
* Make this a user-configurable & importable/exportable option (to choose either the classic hosts-like exact domain matching or this extensive matching, to choose the number of parts in the generated second subdomain in addHost(), and coming up with an option name and description comprehensible for a normal user).
* If source list contains www.example.com, also add example.com.
* Implement support for ublock/adblock domain name syntax (that'll require a lot of thought probably because URLs can't be supported, only domain names, and style filters have to be ignored).
* Add tests.

Upfront, I will probably have very slow progress on this, maybe one commit in 6 months or sth like that.

Thanks

julian-klode commented:
I'd consider adding wildcard support as in #33, but I have not found a good solution yet. Note that we should not split strings (especially in the parser), it's causing too many allocations, slowing down stuff enormously (the big rewrite from string split to substring manipulation in the parser in 0.5 made it much faster!).

SvobodaJakub commented:
It is feature-complete from my POV, so it might be a good time to review whether the proposed change is good.

Split is gone and only efficient string operations are used (I looked into http://www.docjar.com/html/api/java/lang/String.java.html while programming it to see how the methods work to make sure there's nothing too demanding like split.) It could be done more efficiently, but readability would suffer considerably, so that's a balance and I'm willing to shift it in light of feedback/discussion.

As for wildcards, this pull request does provide meaningful wildcards - see the explanation in https://github.com/julian-klode/dns66/pull/185/files#diff-cd52c0f0ecd91c496846307cb1938d1bR270. Why this solution:

* wildcards in a domain name anywhere else than before the domain name are seldom used and would be inefficient
* wildcards after a domain name (in the path part of a url) belong to a different protocol (HTTP as opposed to DNS) and different layer (L7 as opposed to L3) so are inapplicable
* wildcards inside an individual component of a domain name (e.g. "e*ample.com") are seldom, if ever, used, and are not supported
* wildcards in the left side before a dot in the domain name are often used and useful (e.g. *.example.com) - and such wildcards are assumed by this pull request
* the widely used resource-blocking extensions for web browsers implicitly assume a "*." wildcard in front of each domain name if given a plain list of domains to block - e.g. when a blacklist contains "example.com", then also "img.example.com" and "www.example.com" are blocked
* the widely used blacklists do not explicitly use the wildcard characters, it is assumed that the resource-blocking extension in the web browser will also block all subdomains (such assumption is not common for hosts files because this is not how /etc/hosts works, though)
* hence, this pull request doesn't support the wildcard characters and just blocks subdomains (up to the 10th level - capped for worst-case performance reasons) of the blacklisted domains
* also, this pull request also assumes that if "www.example.com" is in a blocklist, then the user wants to also block "example.com" and "whatever.example.com", etc
* all this wildcard-like behavior is called Extensive matching in the UI and can be disabled

Also, this pull request adds support for blacklists in the AdBlock Plus-like syntax. From such files, only the domain-based rules are used; all other rules (CSS-based, url-based) are ignored because they can't be meaningfully blocked on the DNS level.

@julian-klode (and anyone else interested), would you be so kind as to review it, please?

Still WIP, I'm going to dogfood this for several more weeks/months. I'm using my fork constantly, every day, on two android devices, and the last commits will be no exception. It _seems_ fast and functioning well to me. 

Thank you.

jonathanunderwood commented:
If this is feature complete, it's probably worth removing the "WIP" from the PR subject.

SvobodaJakub commented:
It is feature complete except for one breaking detail:

https://github.com/julian-klode/dns66/pull/185/files#diff-cd52c0f0ecd91c496846307cb1938d1bR372

The whitelisting in dns66 works by _removing_ hosts from the blacklist. However, my _extensive filtering_ causes the blacklist to change (e.g. changes `asdf.asdf.example.com` into `example.com` and only that is put into the blacklist). If a whitelist contains `asdf.asdf.example.com`, there is nothing to remove from the blacklist and therefore nothing to whitelist. That's non-obvious, undocumented, and counter-intuitive. So I'm still leaving it as WIP until I solve this problem and make whitelisting work in an intuitive and expected way. Maybe I'll do it over the holidays.

SvobodaJakub committed:
Extensive domain name matching
 - an option to assume wildcards in front of every host name.
https://github.com/julian-klode/dns66/pull/185

SvobodaJakub committed:
Improved tests

SvobodaJakub committed:
Better ABP handling. Fail securely on too long subdomains when extensive matching enabled.

SvobodaJakub committed:
fixed typo

SvobodaJakub committed:
Blocking of pathological domains fixed. It had a bug and blocked everything not in a whitelist.

julian-klode commented:
As I mentioned in August, this is out of scope, and just wrong and can cause a lot of problems, so I'm closing this.

SvobodaJakub commented:
@julian-klode I respect your decision, it is your project after all.

Could you perhaps at least elaborate more on your reasoning, please?

Reasons of my confusion:

- I addressed your feedback from August,
- the feature is purely optional (there is a toggle in the settings),
- I explained why https://github.com/julian-klode/dns66/issues/33 might not be a good fit for dns66 and yet achieved its intended purpose,
- and I don't see how it is `just wrong or can cause a lot of problems`.

That being said, I thank you and the contributors for dns66, it is a wonderful app.
